go/vt/vtctl/reparentutil/durability.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reparentutil
+
+import (
+	"fmt"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/vtctl/reparentutil/promotionrule"
+)
+
+//=======================================================================
+
+// A newDurabler is a function that creates a new durabler based on the
+// properties specified in the input map. Every durabler must
+// register a newDurabler function.
+type newDurabler func(map[string]string) durabler
+
+var (
+	// durabilityPolicies is a map that stores the functions needed to create a new durabler
+	durabilityPolicies = make(map[string]newDurabler)
+	// curDurabilityPolicy is the current durability policy in use
+	curDurabilityPolicy durabler
+)
+
+func init() {
+	// register all the durability rules with their functions to create them
+	registerDurability("none", func(map[string]string) durabler {
+		return &durabilityNone{}
+	})
+	registerDurability("semi_sync", func(map[string]string) durabler {
+		return &durabilitySemiSync{}
+	})
+	registerDurability("cross_cell", func(map[string]string) durabler {
+		return &durabilityCrossCell{}
+	})
+	registerDurability("specified", newDurabilitySpecified)
+}
+
+// durabler is the interface which is used to get the promotion rules for candidates and the semi sync setup
+type durabler interface {
+	promotionRule(*topodatapb.Tablet) promotionrule.CandidatePromotionRule
+	primarySemiSync(*topodatapb.Tablet) int
+	replicaSemiSync(primary, replica *topodatapb.Tablet) bool
+}
+
+func registerDurability(name string, newDurablerFunc newDurabler) {
+	if durabilityPolicies[name] != nil {
+		log.Fatalf("durability policy %v already registered", name)
+	}
+	durabilityPolicies[name] = newDurablerFunc
+}
+
+//=======================================================================
+
+// SetDurabilityPolicy is used to set the durability policy from the registered policies
+func SetDurabilityPolicy(name string, durabilityParams map[string]string) error {
+	newDurabilityCreationFunc, found := durabilityPolicies[name]
+	if !found {
+		return fmt.Errorf("durability policy %v not found", name)
+	}
+	log.Infof("Durability setting: %v", name)
+	curDurabilityPolicy = newDurabilityCreationFunc(durabilityParams)
```

Comment:
are we ok that this is not threadsafe?

Reply:
Well SetDurabilityPolicy is only supposed to be called once from the main thread during initialisation before anyone reads it. In vtctl and vtctld this function is called before we start executing any command. In vtOrc we call it once when we start continuous discovery. After SetDurabilityPolicy has been called, the curDurabilityPolicy should only be read and never written. At this point changing the durability policy requires a restart of the binary. If we add a RPC call to change the durability policy for a binary without requiring a restart we can add a mutex protecting the variable then. If you want I can add it now too. I don't have strong opinions about this

Reply:
Let's add a mutex now, so that if/when we eventually support dynamic durability changes, we are protected.